Protocol errors & Improved handling of failing connection attempts

antonfirsov opened the pull request.

antonfirsov committed:
HttpProtocolException

antonfirsov committed:
Improved handling of failing connection attempts

antonfirsov committed:
wording

CarnaViire commented on HttpProtocolException.md:
```diff
@@ -0,0 +1,29 @@
+
+HTTP2/2 and HTTP/3 defines protocol-level error codes in [RFC 7540 section 7](https://www.rfc-editor.org/rfc/rfc7540#section-7) and [RFC 9114 section 8.1](https://www.rfc-editor.org/rfc/rfc9114.html#section-8.1) (not to be confused with HTTP status codes!). Although such low-level error information is unimportant for most `HttpClient` users, it might be helpful in advanced HTTP/2 or HTTP/3 scenar ...
+
+This is now possible by filtering for [`HttpProtocolException`](https://learn.microsoft.com/en-us/dotnet/api/system.net.http.httpprotocolexception) when calling `HttpClient`:
```
It feels like one more sentence is missing before that one, that would connect the first paragraph with the examples. It could be something like:
"The information about protocol-level error codes is now available in `HttpProtocolException.ErrorCode` property. `HttpProtocolException` would be either an inner exception of `HttpRequestException` (when calling `HttpClient`) or thrown directly (when reading from `HttpContent`'s response stream)"

CarnaViire commented on HttpProtocolException.md:
```diff
@@ -0,0 +1,29 @@
+# Detect HTTP/2.0 and HTTP/3.0 protocol errors
+
+HTTP2/2 and HTTP/3 defines protocol-level error codes in [RFC 7540 section 7](https://www.rfc-editor.org/rfc/rfc7540#section-7) and [RFC 9114 section 8.1](https://www.rfc-editor.org/rfc/rfc9114.html#section-8.1) (not to be confused with HTTP status codes!). Although such low-level error information is unimportant for most `HttpClient` users, it might be helpful in advanced HTTP/2 or HTTP/3 scenar ...
```
- `HTTP2/2` -> `HTTP/2`
- Should we link to grpc-dotnet?
- I feel like we should give examples of protocol-level error codes, when we say that these are not to be confused with status codes, so it would be more obvious to the reader: "e.g. HTTP/2's `REFUSED_STREAM (0x7)` and HTTP/3's `H3_EXCESSIVE_LOAD (0x0107)`" -- or whatever error codes are more useful

CarnaViire commented on HttpProtocolException.md:
````diff
@@ -0,0 +1,29 @@
+Or by catching it directly when working with `HttpContent`'s response stream:
+
+```csharp
+try
````
I would also add
```c#
var responseStream = await response.Content.ReadAsStreamAsync();
```

ManickaP replied:
With using ;-) 

Btw, I'd go even further and start from httpClient.GetAsync.

CarnaViire commented on HttpProtocolException.md:
````diff
@@ -0,0 +1,29 @@
+```csharp
+try
+{
+    var response = await myHttpClientInstance.GetStringAsync(".");
````
let's either put a meaningful url here or use a variable name? `.GetStringAsync(url);`

ManickaP replied:
```suggestion
    var response = await httpClient.GetStringAsync(".");
```
This is subjective, but I really don't like using "my" in code examples. I've seen it copied into production code too many times to be wary of it :'(

CarnaViire commented on Connections.md:
```diff
@@ -0,0 +1,10 @@
+In .NET 7.0 we implemented the following changes to address these issues:
+
+- A failing connection attempt can only fail it's initiating request, and never an unrelated one. If the original request has been handled by the time a connection fails, the connection failure is ignored ([dotnet/runtime#62935](https://github.com/dotnet/runtime/pull/62935)).
+- If a request initiates a new connection, but then becomes handled by another connection from the pool, the new pending connection attempt will time out after a short period, regardless of `ConnectTimeout`. With this change, stalled connections will not stall unrelated requests ([dotnet/runtime#71785](https://github.com/dotnet/runtime/pull/71785)).
```
Let's explicitly mention that in that case the exception would also not reach user code?

And maybe, I don't know how important is that, but mention that if you want to be able to see whether and when that happens, you could use networking telemetry (and link to Miha's cool docs article?)

antonfirsov replied:
I'm not sure if it's possible to deduct the amount of such background failures, and if yes how. The counters I see in the doc only report extablished/existing connections:

https://learn.microsoft.com/en-us/dotnet/fundamentals/networking/networking-telemetry#event-counters-metrics

/cc @MihaZupan

MihaZupan replied:
Looks like we log a random event [when that happens](https://github.com/dotnet/runtime/blob/b49fcc6c5a131608813dcd68760270c18ebd11bb/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs#L1770), but that's only for the the "internal diagnostics".

A nicer approach may be to just listen to [`Sockets.OnConnectFailed`](https://github.com/microsoft/reverse-proxy/blob/d2e47e21798c16d9c105a4fc00cf4d93a4df0f23/src/TelemetryConsumption/Sockets/ISocketsTelemetryConsumer.cs#L33) in general.

CarnaViire commented on Connections.md:
```diff
@@ -0,0 +1,10 @@
+
+In versions prior to .NET 6, in case there is no connection immediately available in the connection pool, a new HTTP request will always spin up a new connection attempt and wait for it. The downside to this is, if it takes a while to establish that connection and another connection becomes available in the meantime, that request would continue waiting for the connection it spawned, hurting laten ...
+
+Unfortunately, the .NET 6.0 implementation turned out to be problematic for some users: a failing connection attempt also fails the request on the top of the request queue, which  may lead to unexpected request failures in [certain scenarios](https://github.com/dotnet/runtime/issues/60654#issue-1030717005). Moreover, if there is a pending connection in the pool that will never become usable eg. b ...
```
I feel like we could elaborate on the second problem a little bit, mention that request cancellation didn't affect connection's creation and one stalled connection could not only "fail new requests unexpectedly" but even lead to massive request timeouts, in case request timeouts were significantly smaller than connect timeouts (especially on Linux)

antonfirsov replied:
> request cancellation didn't affect connection's creation 

I don't think this point is *that* relevant from user perspective.

ManickaP commented on HttpProtocolException.md:
```diff
@@ -0,0 +1,29 @@
+# Detect HTTP/2.0 and HTTP/3.0 protocol errors
```
```suggestion
# HTTP/2 and HTTP/3 Protocol Errors
```
Maybe? I'm a sucker for short headings :-D

ManickaP commented on Connections.md:
```diff
@@ -0,0 +1,10 @@
+# Improved handling of failing connection attempts
```
```suggestion
# Improved handling of connection attempt failures
```
Maybe? Not sure if it's better.

antonfirsov replied:
Renamed it, but I'm not sure either. This title doesn't tell that we are talking about background pending connections. Will see if I can think of a better one.

ManickaP commented on Connections.md:
```diff
@@ -0,0 +1,10 @@
+# Improved handling of failing connection attempts
+
+In versions prior to .NET 6, in case there is no connection immediately available in the connection pool, a new HTTP request will always spin up a new connection attempt and wait for it. The downside to this is, if it takes a while to establish that connection and another connection becomes available in the meantime, that request would continue waiting for the connection it spawned, hurting laten ...
```
I'd start with NET 6 behavior, since that's what caused the issues and what we're mitigating. Explaining 5.0 behavior is fine, but I wouldn't put it as the first thing.

antonfirsov replied:
I think most readers are unaware of the changes in previous versions. It's very hard to clearly and unambiguously explain the story without mentioning the old behavior. In general, to me it feels better to keep the historical context here.

antonfirsov committed:
HttpProtocolException feedback

antonfirsov committed:
connections feedback

antonfirsov commented:
@CarnaViire addressed most of the feedback, or left comments on the previous review where I was reluctant to do so. Happy to push further fixes, will handle it with priority on Thursday.
